Publish the "Caching with Rails" guide

robin850 opened the pull request:
Hello!

This pull request is a potential work in progress in order to publish the caching guide as pull request #21015 intended to do.

I open this pull request to get some feedback and see what's still left to publish this guide. I take the liberty to ping you @ddh and @kaspth as you improved this area a lot recently.

Commit 9065d60 mostly duplicate the `#cache` method's documentation as it contains pretty important stuff but I'm not sure whether this is a good idea.

A possible addition could also be done once #18875 is closed.

Have a nice day!

robin850 committed:
Tiny edits to the "Caching with Rails" guide

* Fix a few typos
* Remove reference to the old `memcache-client` gem
* Remove the "ActiveSupport::Cache::EhCacheStore" part from the guide as
  the gem doesn't seem to be maintained anymore.
* Move the "Custom Cache Stores" part under the "AS::Cache::Store" part
  as they are pretty related.

[ci skip]

robin850 committed:
Add a "Managing dependencies" part to the caching guide [ci skip]

tinogomes committed:
Publish the "Caching with Rails" guide [ci skip]

kaspth commented on guides/source/caching_with_rails.md:
````diff
@@ -152,6 +152,96 @@ With `touch` set to true, any action which changes `updated_at` for a game
+```
+
+As for collection caching, if the partial template doesn't start with a clean
+cache call as mentioned above, you can still benefit from collection caching by
````
I feel like "as mentioned above" is stretching it a bit :smile:

I feel like the parts about collection caching don't really fit in with the other parts about managing dependencies. I think it'd do better in its own section.

kaspth commented on guides/source/caching_with_rails.md:
```diff
@@ -241,6 +332,19 @@ There are some common options used by all cache implementations. These can be pa
+#### Custom Cache Stores
+
+You can create your own custom cache store by simply extending
+`ActiveSupport::Cache::Store` and implementing the appropriate methods. This way,
```
As reader I'm wondering "what are the appropriate methods?" And since I can't see them here, I wonder where I can find out what they are. Lets help the readers out and not cause a mental drip :smile:

robin850 replied:
Yes exactly, that's why this part has been moved under the `ActiveSupport::Cache::Store` section so the appropriate methods are listed just above.

kaspth replied:
Duh! I totally missed that, :smile:. All good :metal:

kaspth commented:
I'm not sure what else is missing, but I've left you some comments now :smile:

I'm definitely :+1: on getting this published.

robin850 commented:
Thank you very much for your quick review! :heart:

robin850 committed:
Add a section about "Collection caching" [ci skip]

robin850 commented:
@kaspth : Okay thanks, updated! :-)

Some comments aren't there anymore as they were inside the commit but here are the changes:

* s/Cache Digests/ActionView::Digestor.
* s/works/work.
* Created a "Collection caching" part under "Fragment caching".
* Removed "as mentioned above".

claudiob commented:
Do you think we should still include Page caching and Action caching in this guide? 

They have been removed from Rails a while ago.

I think we might be better off focusing on what Rails offers right now out of the box.

claudiob commented:
Additionally, I think the very beginning of this guide could be improved. Right now it says:

> This guide will teach you what you need to know about avoiding that expensive round-trip to your database and returning what you need to return to the web clients in the shortest time possible.
>
> After reading this guide, you will know:
> * Page and action caching (moved to separate gems as of Rails 4).
> * Fragment caching.
> * Alternative cache stores.
> * Conditional GET support.

Since this is a guide, I think we can take a wider approach and introduce what caching is and why is beneficial. For instance:

> Caching means to store the content generated by a request and to reuse it when responding to similar requests.
 
> Caching is arguably one of the hardest problems in computer science.
> At the same time, it is the most effective way to boost the performance of an application. 
> Through caching, web sites running on a single server with a single database can sustain a load of thousands of concurrent users.
 
> Rails provides a set of caching methods out of the box.
> This guide will teach you the scope and purpose of each one of them.
> Master these techniques and you will be able to scale your Rails applications, serving million of views without incurring in exorbitant delays or server bills.

claudiob committed:
Improve the "Caching with Rails" guide's introduction [ci skip]

robin850 commented:
> Do you think we should still include Page caching and Action caching in this guide? 

I vaguely remember that I asked this on Campfire once and IIRC, we should keep it in the guide ; it's extracted from Rails core but it is still maintained. At least, we aren't going in depth &mdash; just mentioning that these strategies exist.

> I think we can take a wider approach and introduce what caching is and why is beneficial

Good call! :+1: Commit 69e97ea08f7e52c47a5fc9c4b7ef1a3bd714a5ec adds that ; I took the liberty to set you as an author.

Thank you very much for your feedback!

kaspth commented on guides/source/caching_with_rails.md:
```diff
@@ -3,12 +3,25 @@
-This guide is an introduction to speeding up your Rails app with caching.
+This guide is an introduction to speeding up your Rails application with caching.
+
+Caching means to store the content generated during the request-response cycle
```
"the content" makes it feel like it's the whole response body, instead of merely part of it (as it could be with fragment caching). I would `the` before content.

kaspth commented on guides/source/caching_with_rails.md:
```diff
@@ -3,12 +3,25 @@
+Caching means to store the content generated during the request-response cycle
+and to reuse it when responding to similar requests.
+
+Caching is arguably one of the hardest problems in computer science. At the same
```
`computer science` sounds too abstract for me. That's hardly what most web developers do in day to day development.

I think it would be better to just remove the first sentence and go with:

```
Caching is often the most effective way to boost an application's performance.
```

kaspth commented on guides/source/caching_with_rails.md:
```diff
@@ -3,12 +3,25 @@
+Rails provides a set of caching features out of the box. This guide will teach
+you the scope and purpose of each one of them. Master these techniques and you
+will be able to scale your Rails applications, serving million of views without
+incurring in exorbitant delays or server bills.
```
:scissors: `in` after `incurring`.

kaspth commented on guides/source/caching_with_rails.md:
```diff
@@ -3,12 +3,25 @@
+
+Rails provides a set of caching features out of the box. This guide will teach
+you the scope and purpose of each one of them. Master these techniques and you
+will be able to scale your Rails applications, serving million of views without
```
`millions` should be plural.

I'd also think about rephrasing to:

```ruby
Master these techniques and your Rails applications can serve millions of views without exorbitant response times or server bills.
```

kaspth commented:
Great suggestions, @claudiob :smile:

@robin850 what the latest changes this is starting to feel really good :+1:

robin850 commented:
@kaspth : Thanks, updated! :-)

kaspth commented on guides/source/caching_with_rails.md:
```diff
@@ -103,6 +115,29 @@ If you want to cache a fragment under certain conditions, you can use
+The `render` helper allows you to do caching for a collection so that it can
+automatically use any cached renders for a template by reading them at once
+instead of one by one like the previous example with `each` did. For instance,
+for a collection of products:
```
I'd like to suggest a revision of this. What do you think? :smile:

```
The `render` helper can also cache individual templates rendered for a collection. It can even one up the previous example with `each` by reading all cache templates at once instead of one by one. This is done automatically if the template rendered by the collection includes `cache` call. Take a collection that renders a `products/_product.html.erb` partial for each element:

render products

If `products/_product.html.erb` starts with a `cache` call like so:

<% cache product do %>
  <%= product.name %>
<% end %>

All the cached templates from previous renders will be fetched at once with much greater speed. There's more info on how to make your templates [eligible for collection caching](http://api.rubyonrails.org/classes/ActionView/Template/Handlers/ERB.html#method-i-resource_cache_call_pattern).
```

robin850 commented:
@kaspth : Much better, thanks! :+1: I didn't even know the expression "to one up"!

kaspth commented:
Alright, this feels good to me :+1:

robin850 commented:
Yay, thanks! \o/
